sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -327,121 +329,314 @@ bool WinHttpTransport::VerifyCertificatesInChain(
   return true;
 }
 
-/**
- * Called by WinHTTP when sending a request to the server. This callback allows us to inspect the
- * TLS certificate before sending it to the server.
- */
-void WinHttpTransport::StatusCallback(
-    HINTERNET hInternet,
-    DWORD_PTR dwContext,
-    DWORD dwInternetStatus,
-    LPVOID,
-    DWORD) noexcept
+namespace {
+
+#define APPEND_ENUM_STRING(id) \
+  if (internetStatus & (id)) \
+  { \
+    rv += std::string(#id) + " "; \
+  }
```

Comment:
We try to avoid macros. Consider updating to a named (inline) function.

https://azure.github.io/azure-sdk/cpp_implementation.html#macros

Reply:
Named functions don't have the tokenizing operator (`#`).  The macro depends on that functionality.

Reply:
@CaseyCarter is there a way to write this implementation in a named-function (relatively cleanly), which doesn't rely on the stringizing operator which is only available in a macro?
https://learn.microsoft.com/en-us/cpp/preprocessor/stringizing-operator-hash?view=msvc-170

If not, we may need to update the guidelines, because this use-case of a macro isn't listed as one of the allowed exceptions in our guidelines.

Reply:
This macro is intended for diagnostic purposes, which I believe makes it squarely within the guidelines. It is not used as a part of any public API contract.

Reply:
Hmm, I don't think that was the intent of that guideline, but we can make it fit :)

@BillyONeal, since you wrote these guidelines initially (here: https://github.com/Azure/azure-sdk/pull/1146),  can you expand on this a bit? What do you mean by emission/suppression of diagnostic.

Btw Larry, the "Implementation" page of the guidelines are in fact about the Implementation and not just the public API contract. So, there are parts of it that mention coding patterns to adhere to, that is only implementation detail.

P.S. Unresolving only to make the conversation easier to find for others (not to block the merge), feel free to resolve and merge upon response/approval.

Reply:
github hides this conversation even if it's unresolved.

Reply:
I understand that heavy use of macros can be challenging, especially macros which hide significant amounts of code. And it can be tricky to write macros correctly. However, macros can also be used to great effect, especially when emitting diagnostics.

Reply:
I think I meant things like emission meaning `UNREFERENCED_PARAMETER` or similar, and suppression being warning disable pragmas or similar? It's been too long. This was also early enough that such guidelines were "if you want to not do this lets talk about it and we might make more exceptions" that sort of thing, so I think you should review this without regard for what the guidelines said.

If you do keep this I recommend doing:

```
rv += #id " "
```

instead, assuming `rv` is intended to be a `std::string` or similar. (That is, just do the concat inside the string literal rather than needing to construct a temporary std::string to do that)

Reply:
@BillyONeal  D'oh, of course that's better - for some reason, I continually forget that I don't need to concatenate adjacent string constants.